### PR TITLE
Compatibility of Blinks and Actions

### PR DESCRIPTION
**Pull request by tsmbl**
**Context & Problem**

Based on sRFC https://forum.solana.com/t/srfc-31-compatibility-of-blinks-and-actions/1892

> Response headers are proposed to be set by Action developers:
- `X-Action-Version` to show what spec version the action API server is using
- `X-Blockchain-Ids` to list blockchains the action supports.
> In the future request headers are proposed to be set by Blink clients:
- `X-Accept-Action-Version` to show the max spec version the Blink client supports.
- `X-Accept-Blockchain-Ids` to list blockchains the client supports.

Response headers should be used to decide if Action supported by client.

**Solution**

1. Allow specifying action adapter metadata, this data can be used in default supportability checks and later can be used to set X-Accept-Blockchain-Ids` request header.
2. Parametrize action with `supportedBlockchainIds` strategy  
3. Provide utility functions for supportability checks by comparing action & blink metadata:
- Semver versions: action is not supported if action provider spec version > blink library spec version, patch version is ignored during comparison.
- Blockchain IDs:  action is not supported if there's at least one blockchain id that is not supported by blink library client
4. Provide default implementation for metadata and supportability in `ActionConfig` 

**Breaking changes**
- move `isSupported` from `ActionAdapter`, to `Action` and to  `ObserverOptions` correspondingly

**Example adapter initialization**
1. Default settings in `ActionConfig`: in this case client supports solana mainnet & devnet + spec version from `package.json`

```typescript
const defaultActionConfig: ActionConfig = new ActionConfig(
  'http://localhost:8080',
  {
    connect: async (c) => {
      return 'connected';
    },
    signTransaction: async (tx, c) => {
      return { signature: 'signed' };
    },
  },
);

```
2. Configure metadata in `ActionConfig`: in this case client supports `supportedBlockchainIds` + spec version from package.json, `supportedBlockchainIds` can be later used to set `X-Accept-Blockchain-Ids` request header
```typescript
const actionConfigWithMetadata: ActionConfig = new ActionConfig(
  'http://localhost:8080',
  {
    metadata: {
      supportedBlockchainIds: [BlockchainIds.SOLANA_MAINNET],
    },
    connect: async (c) => {
      return 'connected';
    },
    signTransaction: async (tx, c) => {
      return { signature: 'signed' };
    }
  },
);

```
3. Overriding default `supportStrategy` in `useAction`
```typescript
 const { action } = useAction({
    supportStrategy: async () => ({
      isSupported: false,
      message: 'Message to be displayed.',
    }),
    // ...rest
  });
```

4. Overriding default `supportStrategy` in `setupTwitterObserver`

```typescript
setupTwitterObserver(
  null as unknown as ActionAdapter,
  {},
  {
    supportStrategy: async (a) => {
      const versionSupported = a.metadata.version === '2.2.8';
      if (!versionSupported) {
        return {
          isSupported: false,
          message: 'Version not supported',
        };
      }
      return {
        isSupported: true,
      };
    },
  },
);
```

Related PR in actions spec: https://github.com/solana-developers/solana-actions/pull/21
Related PR in actions-proxy: https://github.com/dialectlabs/actions-backend/pull/2

### CODE REVIEW

File: src/api/Action/action-supportability.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+/**
+ * Max spec version the Blink client supports.
+ */
+export const MAX_SUPPORTED_ACTION_VERSION = ACTIONS_SPEC_VERSION;
+/**
+ * Baseline action version to be used when not set by action provider.
+ * Defaults to latest release that doesn't support versioning.
+ */
+export const BASELINE_ACTION_VERSION = '2.0.0';
```
Comment (zhelezkov) [inline src/api/Action/action-supportability.ts:13]: why 2.0.0? shouldn't it be 2.2 or smth?

File: src/api/Action/action-supportability.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * Default implementation for checking if an action is supported.
+ * Checks if the action version and the action blockchain IDs are supported by blink.
+ * @param context Action context.
+ * @param supportedBlockchainIds List of CAIP-2 {@link BlockchainIds} the client supports.
+ *
+ * @see {isVersionSupported}
+ * @see {isBlockchainSupported}
+ */
+export function defaultCheckSupported(
+  context: Omit<ActionContext, 'triggeredLinkedAction'>,
```
Comment (zhelezkov) [inline src/api/Action/action-supportability.ts:44]: mb like?
```suggestion
  context: Pick<ActionContext, 'action'>,
```

File: src/api/Action/Action.ts [hunk 3]
```diff
@@ -25,7 +30,7 @@ export class Action {
   private constructor(
     private readonly _url: string,
     private readonly _data: ActionGetResponse,
-    private readonly _metadata: ActionMetadata,
+    private readonly _metadata: Partial<ActionMetadata>,
```
Comment (zhelezkov) [inline src/api/Action/Action.ts:33]: shouldn't we change `ActionMetadata` type itself, to avoid using `Partial` later?

**Review by zhelezkov — APPROVED** [review]
this is really good job
Love the documented code and tests!

File: src/ui/ActionContainer.tsx [hunk 4]
```diff
@@ -266,6 +289,28 @@ export const ActionContainer = ({
...
+        setSupportability(supportability);
+      } finally {
+        dispatch({ type: ExecutionType.RESET });
+      }
+    };
+
+    checkSupportability(action);
+  }, [action]);
+
+  // const supportability: ActionSupportability = useMemo(() => {
```
Comment (fsher) [inline src/ui/ActionContainer.tsx:309]: minor: remove commented code pls

**Review by fsher — APPROVED** [review]
amazing job! minor comments

File: src/api/Action/Action.ts [hunk 5]
```diff
@@ -118,15 +135,26 @@ export class Action {
     this._adapter = adapter;
   }
 
+  public isSupported() {
```
[the remark this answers is not in the document]
Comment (tsmbl) [reply src/api/Action/Action.ts:138]: how do you propose to handle errors during strategy calculation? just assume action is not supported?

**Review by fsher — APPROVED** [review]

**Pull request merged**